src/arch/riscv64/interrupts.rs

```diff
@@ -0,0 +1,193 @@
+use crate::println;
+use crate::utils;
+use core::mem::size_of;
+
+use riscv;
+
+const PLIC_ADDR: usize = 0xc000000;
+const PLIC_ENABLE_OFFSET: usize = 0x002000;
+const UART_IRQ_NB: usize = 10;
+
+#[derive(Debug)]
+pub enum InterruptsState {
+    DISABLE = 0,
+    ENABLE = 1,
+}
+
+pub fn state() -> InterruptsState {
+    match riscv::register::sstatus::read().sie() {
+        false => InterruptsState::DISABLE,
+        true => InterruptsState::ENABLE,
+    }
+}
+
+fn set(status: InterruptsState) {
+    match status {
+        InterruptsState::DISABLE => unsafe { riscv::register::sstatus::clear_sie() },
+        InterruptsState::ENABLE => unsafe { riscv::register::sstatus::set_sie() },
+    }
+}
+
+pub fn init() {
+    set(InterruptsState::DISABLE);
+
+    // Enable only Supervisor-level external interrupts
+    unsafe {
+        riscv::register::sie::clear_usoft();
+        riscv::register::sie::set_ssoft();
+        riscv::register::sie::clear_utimer();
+        riscv::register::sie::set_stimer();
+        riscv::register::sie::clear_uext();
+        riscv::register::sie::set_sext();
+    }
+
+    let test_addr = unsafe { stvec_base as usize };
+    println!("Vector base addr: ${:x}", test_addr);
+    unsafe {
+        riscv::register::stvec::write(test_addr, riscv::register::stvec::TrapMode::Direct);
+    }
+
+    //TODO: Setup PLIC?
+    // let plic_uart_priority = (PLIC_ADDR + 4 * UART_IRQ_NB) as *mut u32;
+    // let plic_uart_enable = (PLIC_ADDR + PLIC_ENABLE_OFFSET + UART_IRQ_NB / 32) as *mut u32;
+    // unsafe {
+    //     *plic_uart_priority = 1;
+    //     *plic_uart_enable = 1 << 10;
+    // }
+
+    set(InterruptsState::ENABLE);
+
+    // let v = 0x0 as *mut u64;
+    // unsafe {
+    //     v.write_volatile(0);
+    // }
+    //
+    // println!("We can't be here");
+}
+
+extern "C" {
+    fn stvec_base();
+}
+
+#[naked]
+#[no_mangle]
+/// This function will never get called. It's only a wrapper around stvec_base which is aligned to
+/// 4 as needed for the base field in stvec register
+pub extern "C" fn stvec_base_wrapper() {
+    unsafe {
+        // As rust does not provide a way to have align a function
+        // [yet](https://github.com/rust-lang/rust/issues/75072), we do it here and use
+        // stvec_base address in stvec register
```

Comment:
This will not appear in the documentation because of the usage of `//` instead of `///`. I suggest putting it above the function, with the remaining documentation

Reply:
Ok